### PR TITLE
Add database migrations for AO/CD roles

### PR DESCRIPTION
**Pull request by ashley-weaver**
## 🎫 Ticket

https://jira.cms.gov/browse/DPC-3638

## 🛠 Changes

(What was added, updated, or removed in this PR.)

## ℹ️ Context for reviewers

(Background context, more in-depth details of the implementation, and anything else you'd like to call out or ask reviewers.)

## ✅ Acceptance Validation

(How were the changes verified? Did you fully test the acceptance criteria in the ticket? Provide reproducible testing instructions and screenshots if applicable.)

## 🔒 Security Implications

- [ ] This PR adds a new software dependency or dependencies.
- [ ] This PR modifies or invalidates one or more of our security controls.
- [ ] This PR stores or transmits data that was not stored or transmitted before.
- [ ] This PR requires additional review of its security implications for other reasons.

If any security implications apply, add Jason Ashbaugh (GitHub username: StewGoin) as a reviewer and do not merge this PR without his approval.

### CODE REVIEW

File: dpc-portal/db/migrate/20240307212048_create_ao_org_links.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  def change
+    create_table :ao_org_links do |t|
+      t.integer :user_id, null: false
+      t.integer :provider_organization_id, null: false
+      t.timestamps null: false
+    end
+
+    add_foreign_key :ao_org_links, :users, :user_id
+    add_foreign_key :ao_org_links, :provider_organizations, :provider_organization_id
+    add_index :ao_org_links, :user_id, unique: true
```
Comment (jdettmannnava) [inline dpc-portal/db/migrate/20240307212048_create_ao_org_links.rb:11]: not unique! A user might be ao to a bunch of orgs! Need uniqueness on combination of the two

File: dpc-portal/db/migrate/20240307212048_create_ao_org_links.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class CreateAoOrgLinks < ActiveRecord::Migration[7.1]
+  def change
+    create_table :ao_org_links do |t|
+      t.integer :user_id, null: false
+      t.integer :provider_organization_id, null: false
+      t.timestamps null: false
+    end
+
+    add_foreign_key :ao_org_links, :users, :user_id
+    add_foreign_key :ao_org_links, :provider_organizations, :provider_organization_id
```
Comment (jdettmannnava) [inline dpc-portal/db/migrate/20240307212048_create_ao_org_links.rb:10]: When they are the same, you don't need to give the attribute.
So `add_foreign_key :ao_org_links, :provider_organizations` is enough

File: dpc-portal/spec/factories/cd_org_links.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+FactoryBot.define do
+  factory :cd_org_link do
+    user_id { 1 }
+    provider_organization_id { 1 }
+    invitation_id { 1 }
+    disabled_at { nil }
```
Comment (jdettmannnava) [inline dpc-portal/spec/factories/cd_org_links.rb:6]: not necessary

File: dpc-portal/app/models/ao_org_link.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+class AoOrgLink < ApplicationRecord
+  attr_reader :user_id, :provider_organization_id
```
Comment (jdettmannnava) [inline dpc-portal/app/models/ao_org_link.rb:2]: You will need associations with the objects via belongs_to

File: dpc-portal/app/models/cd_org_link.rb [hunk 5]
```diff
@@ -1,26 +1,7 @@
-# frozen_string_literal: true
+class CdOrgLink < ApplicationRecord
+  attr_reader :user_id, :provider_organization_id, :invitation_id
```
Comment (jdettmannnava) [inline dpc-portal/app/models/cd_org_link.rb:2]: You will need associations with the objects via belongs_to

File: dpc-portal/app/models/cd_org_link.rb [hunk 6]
```diff
@@ -1,26 +1,14 @@
 # frozen_string_literal: true
 
-# Placeholder for eventual ActiveRecord for saving this
-class CdOrgLink
-  attr_reader :given_name, :family_name, :email, :pending, :verification_code, :activated_at
+# Link between credential delegate and provider organization
+class CdOrgLink < ApplicationRecord
+  attr_reader :user_id, :provider_organization_id, :invitation_id
```
Comment (jdettmannnava) [inline dpc-portal/app/models/cd_org_link.rb:5]: attr_reader is not necessary

File: dpc-portal/spec/components/page/credential_delegate/list_component_spec.rb [hunk 7]
```diff
@@ -86,19 +86,6 @@
         is_expected.to include(normalize_space(expected_html))
       end
 
-      it 'has a row' do
```
Comment (jdettmannnava) [inline dpc-portal/spec/components/page/credential_delegate/list_component_spec.rb:89]: We want to test this.

File: dpc-portal/app/components/page/credential_delegate/list_component_preview.rb [hunk 8]
```diff
@@ -15,10 +15,7 @@ def empty
       end
 
       def active
-        cds = [
-          CdOrgLink.new(given_name: 'Bob', family_name: 'Hodges', email: 'bob@example.com', pending: false),
-          CdOrgLink.new(given_name: 'Lisa', family_name: 'Franklin', email: 'lisa@example.com', pending: false)
-        ]
+        cds = [CdOrgLink.new, CdOrgLink.new]
```
Comment (jdettmannnava) [inline dpc-portal/app/components/page/credential_delegate/list_component_preview.rb:18]: Will want users so have data to show

File: dpc-portal/app/models/cd_org_link.rb [hunk 9]
```diff
@@ -1,26 +1,14 @@
...
-    @email = email
-    @pending = pending
-    @verification_code = verification_code || (Array('A'..'Z') + Array(0..9)).sample(6).join
-    @activated_at = 1.day.ago
-  end
+  belongs_to :user, required: true
+  belongs_to :provider_organization, required: true
+  belongs_to :invitation, required: true
 
   def show_attributes
```
Comment (jdettmannnava) [inline dpc-portal/app/models/cd_org_link.rb:11]: still need the attributes:
{ full_name: "#{user&.given_name} #{user.&family_name}"...

File: dpc-portal/spec/components/page/credential_delegate/list_component_spec.rb [hunk 10]
```diff
@@ -87,12 +90,10 @@
       end
 
       it 'has a row' do
-        activated = 1.day.ago.strftime('%m/%d/%Y at %l:%M%p UTC')
         expected_html = <<~HTML
           <tr>
             <td data-sort-value="Bob Hodges">Bob Hodges</td>
             <td data-sort-value="bob@example.com">bob@example.com</td>
-            <td data-sort-value="#{activated}">#{activated}</td>
```
Comment (jdettmannnava) [inline dpc-portal/spec/components/page/credential_delegate/list_component_spec.rb:95]: I think this still needs to be here. It can be CdOrgLink.created_at

File: dpc-portal/spec/models/ao_org_link_spec.rb [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  let(:provider_organization) { build(:provider_organization) }
+  let(:user) { User.new(given_name: 'Bob', family_name: 'Hodges', email: 'bob@example.com') }
+  let(:ao_org_link) { build(:ao_org_link, user:, provider_organization:) }
+
+  it 'has foreign keys' do
+    expect(ao_org_link.user).to eq user
+    expect(ao_org_link.provider_organization).to eq provider_organization
+  end
+
+  it 'shows attributes' do
```
Comment (jdettmannnava) [inline dpc-portal/spec/models/ao_org_link_spec.rb:15]: I don't think we need this?

Maybe add a test for no duplicate user-provider_org?

File: dpc-portal/spec/models/cd_org_link_spec.rb [hunk 12]
```diff
@@ -3,18 +3,28 @@
...
   end
 
   it 'shows attributes' do
-    cd_org_link = CdOrgLink.new(given_name: 'Bob', family_name: 'Hodges', email: 'bob@example.com', pending: false,
-                                verification_code: 'ABC123')
     attrs = cd_org_link.show_attributes
     expect(attrs['full_name']).to eq 'Bob Hodges'
     expect(attrs['email']).to eq 'bob@example.com'
     expect(attrs['verification_code']).to eq 'ABC123'
-    expect(attrs['activated_at']).to eq 1.day.ago.to_s
```
Comment (jdettmannnava) [inline dpc-portal/spec/models/cd_org_link_spec.rb:18]: Maybe need this

**Review by jdettmannnava — APPROVED** [review]
LGTM
